Sources/Fittings/interpolate.cpp

```diff
@@ -64,11 +75,95 @@ void InterpolateSurface(double* points,
                 pts_v[v * dim + i] = tmp_control_points[(u + (size_u * v)) * dim + i];
             }
         }
-        coefficient_matrix = BuildCoefficientMatrix(degree_v, knot_vector_v, v_l, pts_v, size_v);
+        coefficient_matrix = BuildCoefficientMatrix(degree_v, knot_vector_v, v_l, size_v, size_v);
         tmp_result = LUSolve(coefficient_matrix, pts_v, size_v, dim); 
         std::move(tmp_result.begin(), tmp_result.end(), std::back_inserter(control_points));
     }
     
-    delete[] pts_u, pts_v;
+    delete pts_u, pts_v;
+
+}
+
+
+double ApproximateCurve(double* points,
+                        int& num_points,
+                        int& dim,
+                        int& degree,
+                        int& n_ctps,
+                        std::vector<double>& u_k,
+                        std::vector<double>& knot_vector,
+                        std::vector<double>& coefficient_matrix,
+                        std::vector<double>& control_points) {
+  
+  // Number of variable n_ctps (non-predetermined by P0 and Pm)
+  int n_ctps_v = n_ctps -2;
+
+  // Calculate R_k
+  double* R_points = new double[(num_points - 2) * dim];
+
+  // equation 9.63
+  for (int k{1}; k < num_points - 1; k++){
+    for (int i_dim{0}; i_dim < dim ; i_dim++){
+      R_points[(k - 1) * dim + i_dim] = points[k * dim + i_dim] -
+        coefficient_matrix[k * n_ctps] * points[i_dim] -
+        coefficient_matrix[k * n_ctps + n_ctps -1] * points[(num_points-1) * dim + i_dim];
+    }
+  }
+
+  // Assemble Least Square Problem
+  // Reduced Matrix
+  std::vector<double> LSmatrix;
+  LSmatrix.assign(n_ctps_v * n_ctps_v, 0.0);
+
+  // Reduced Vector NTR
+  double* NtR = new double[n_ctps_v * dim];
+  for (int i{} ; i < n_ctps_v * dim; i++) NtR[i] = 0.0;
+  
+  // Calculate LS-System *symmetrize*
+  for (int i{1}; i < n_ctps - 1; i++) {
+    for (int k{1}; k < num_points - 1; k++){
+      // Assemble Reduced NTN
+      for (int j{1}; j < n_ctps - 1; j++){
+        // Calculate NT N
+        LSmatrix[(i - 1) * n_ctps_v + j - 1] += 
+          coefficient_matrix[k * n_ctps + i] * 
+          coefficient_matrix[k * n_ctps + j];
+      }
+
+      // Calculate NtR
+      for (int i_dim{0}; i_dim < dim; i_dim++){
+        NtR[(i - 1) * dim + i_dim] += coefficient_matrix[k * n_ctps + i] *
+          R_points[(k - 1) * dim + i_dim];
+      }
+    }
+  }
+
+  // Solve linear system with LU-decomposition
+  control_points = LUSolve(LSmatrix, NtR, n_ctps_v, dim);
+  
+  // Add P0 at front and Pm at end (might be inefficient)
+  for (int i{0}; i < dim ; i++){
+    control_points.insert(control_points.begin(), points[dim-i-1]);
+    control_points.push_back(points[(num_points-1) * dim + i]);
+  }
```

Comment:
```suggestion
  // Add P0 at front and Pm at end (might be inefficient)
  control_points.resize(control_points.size() + dim * 2);
  for (int i{n_ctps_v * dim - 1}; i >= 0; i--) {
    control_points[i+dim] = control_points[i];
  }
  for (int i{0}; i < dim ; i++) {
    control_points[i]  = points[i];
    control_points[(n_ctps_v +1) * dim + i] = points[(num_points-1) * dim + i];
  }
```

This is actually more efficient, my bad...